### DISCUSSION TITLE
SD-Card upload

### DISCUSSION THREAD

**Discussion by infamous-panda** [terjeio/ioSender#148] [General]
If the upload button is greyed out for the SD-Card. What could be the reasons?

When there is no SD-Card. It states that there is no card mounted. If the functionality is not in the firmware the TAB would not be there correct?

**Comment by terjeio** [comment]
> If the upload button is greyed out for the SD-Card. What could be the reasons?

The YModem protocol has to be enabled in the controller, I have not documented how as it is a little unstable. Change the SD card enable to `#define SDCARD_ENABLE 2` in _my_machine.h_ if you want to try.

> If the functionality is not in the firmware the TAB would not be there correct?

Correct.

**Reply by infamous-panda** [reply]
My G-Code files are very dense raster engravings and I find USB streaming to be a large bottleneck. 

I just did a quick test by running a program from the SD-card and the resulting simulated speed was 6-8x faster than what I could achieve with USB. I wish this feature could be a little more fleshed out. Specifically

1. Stable YModem ( I did not actually try this, I just physically removed the sd-card and loaded it through windows)
2. Have an option to load the code to show up on the console or at the very least a way to hit cycle start for SD-card programs. 
3. Also I have to physically hit the stop button to end the program. Even when the program completes the timer is still running and the machine is "active"

**Reply by terjeio** [reply]
1. I am not sure I want to spend more time on YModem - I could not figure out why it sometimes refused to start a transfer.
FTP is a better alternative if you have a networking capable controller. The next sender release will support FTP, WinSCP and FileZilla (and surely others) can already be used to maintain the SD card contents if FTP is enabled in the controller.

2. To show up in in the program listing and the 3D viewer the program has to be downloaded first. Another layer of complexity but I guess doable.

3. Which controller and sender versions do you have installed? Is the program terminated with M2 or M30? Cycle start can be used to restart the same program if you right click on the filename and select _Enable rewind_ in the menu when first starting it.

**Reply by infamous-panda** [reply]
1. It's been a while but to access ftp would all that is needed is to type in ftp://192.168.1.199:23? IP and port shown just as an example.

2. This would be great but not at the expense of the speed. At the very least the program should be queued and the cycle start button functional.   

3. I am using the Teensy 4.1 with a current firmware and Sender 2.0.34. For some reason I cannot get 2.0.35 to launch on any of my 3 windows 10 machines. But it launches on my 1 windows 11. I have not added M2 or M30 to any of programs. I usually just add M5 to turn off the laser. I'll give this a try.

**Reply by terjeio** [reply]
1. Type in where? In your web browser? If so mine wants to open an external program. I use WinSCP and I have tested with FileZilla.
Note that FTP needs to be enabled in the controller by uncommenting `#define FTP_ENABLE 1` in _my_machine.h_. You can check enabled status in _Help > About_:

![image](https://user-images.githubusercontent.com/20260062/141527128-06486526-1f48-4ae4-856b-ed807942e0f7.png)

2. I'll consider this, scrolling the program listing may require line numbers - 3D view should work out of the box.

3. Can you try the latest [edge version](http://www.io-engineering.com/downloads)? Do you get any error message when it fails? Anything in the Windows event log?

BTW if you want to press cycle start before starting gcode on the SD-card add `M1` as the first line.

**Reply by infamous-panda** [reply]
> BTW if you want to press cycle start before starting gcode on the SD-card add `M1` as the first line.

No big deal. It would just be nice to be consistent with SD card loading or file explorer from windows. I assume M1 would require pressing cycle start twice if loaded from windows?

**Reply by terjeio** [reply]
>> BTW if you want to press cycle start before starting gcode on the SD-card add M1 as the first line.

> No big deal. It would just be nice to be consistent with SD card loading or file explorer from windows.

I'll have to add this as an option in the controller as I believe it not easy to do in the sender. Let me think about it.

> I assume M1 would require pressing cycle start twice if loaded from windows?

Correct.

**Reply by terjeio** [reply]
Sometimes I forget the flexibility grblHAL has, pausing on SD card execution can be solved by a plugin:

```
#include "grbl/hal.h"
#include "grbl/gcode.h"

static on_stream_changed_ptr on_stream_changed;
static on_report_options_ptr on_report_options;

static void stream_changed (stream_type_t type)
{
    if(on_stream_changed)
        on_stream_changed(type);

    if(type == StreamType_SDCard)
        gc_execute_block("M1");
}

static void report_options (bool newopt)
{
    on_report_options(newopt);

    if(!newopt)
        hal.stream.write("[PLUGIN:SD Pause v0.01]" ASCII_EOL);
}

void my_plugin_init (void)
{
    on_stream_changed = grbl.on_stream_changed;
    grbl.on_stream_changed = stream_changed;

    on_report_options = grbl.on_report_options;
    grbl.on_report_options = report_options;
}
```
This traps the stream switching to the SD card and sends a `M1` to the parser to enter hold state. The `grbl.on_stream_changed` event was added 2021-09-13 so a build date equal or later than that is needed.

Copy the code above to a file named _my_plugin.c_ in the same folder that _driver.c_ is located and recompile/reflash.

**Reply by infamous-panda** [reply]
> Sometimes I forget the flexibility grblHAL has, pausing on SD card execution can be solved by a plugin:
> 
> ```
> #include "grbl/hal.h"
> #include "grbl/gcode.h"
> 
> static on_stream_changed_ptr on_stream_changed;
> static on_report_options_ptr on_report_options;
> 
> static void stream_changed (stream_type_t type)
> {
>     if(on_stream_changed)
>         on_stream_changed(type);
> 
>     if(type == StreamType_SDCard)
>         gc_execute_block("M1");
> }
> 
> static void report_options (bool newopt)
> {
>     on_report_options(newopt);
> 
>     if(!newopt)
>         hal.stream.write("[PLUGIN:SD Pause v0.01]" ASCII_EOL);
> }
> 
> void my_plugin_init (void)
> {
>     on_stream_changed = grbl.on_stream_changed;
>     grbl.on_stream_changed = stream_changed;
> 
>     on_report_options = grbl.on_report_options;
>     grbl.on_report_options = report_options;
> }
> ```
> 
> This traps the stream switching to the SD card and sends a `M1` to the parser to enter hold state. The `grbl.on_stream_changed` event was added 2021-09-13 so a build date equal or later than that is needed.
> 
> Copy the code above to a file named _my_plugin.c_ in the same folder that _driver.c_ is located and recompile/reflash.

Hi Terje,

I am not clear on how to plugin the above code to my_plugin.c. Any which way I have done it fails to compile in Arduino.

**Reply by terjeio** [reply]
I compiles and works for me in Arduino for Teensy.

Create _my_plugin.c_ (not _my_plugin.c.txt_ as sometimes Windows does - it has  the IMO bad habit of hiding file extensions...) in the _src_ folder, copy the code above into it and recompile. If you get compile errors then be sure you have an up to date grblHAL version, if you have post the compiler errors here.

**Comment by infamous-panda** [comment]
I'm getting a the below error with Filezilla. probably missing something simple.  

![image](https://user-images.githubusercontent.com/54411216/141643765-0dac367c-7db1-4ca3-b1e2-dc3bf0492991.png)

**Reply by terjeio** [reply]
Could be FTP is not checked in _Settings: GRBL_? I forgot to mention that this is needed if it is a firmware update.

![image](https://user-images.githubusercontent.com/20260062/141644146-0d7396b3-e2a7-4428-a78b-e15e155dece6.png)

A power recycle of the controller is required after changing this.

**Reply by terjeio** [reply]
BTW FTP uses port 21, 23 is Telnet.

**Comment by infamous-panda** [comment]
2.0.35 or the edge version you linked do not launch on my windows 10 machines for some reason.

**Reply by terjeio** [reply]
Is it a machine with no controller connected? The sender exists silently if there are no COM ports available. I have a pending fix for that, see issue #149.

**Reply by infamous-panda** [reply]
> Is it a machine with no controller connected? The sender exists silently if there are no COM ports available. I have a pending fix for that, see issue #149.

Nah, It's the same machine I am running 2.034. I have only gotten 2.0.35 to launch on my one windows 11 machine..

**Reply by terjeio** [reply]
Is there anything in the Windows event logs that can explain this?

**Reply by infamous-panda** [reply]
> Is there anything in the Windows event logs that can explain this?

First time I looked at windows event viewer! I posted a log on a similar thread under Issues. 
So this happens with both 2.0.35 and the edge version you sent recently.

**Comment by infamous-panda** [comment]
Thanks! that was it.

**Comment by infamous-panda** [comment]
Yes I figured that out. Also works with windows explorer now so no need for
extra programs

**Comment by infamous-panda** [comment]
Do you replace everything or add this to the end of the existing code in
my_plugin.c

On Tue, Nov 16, 2021 at 12:32 PM Terje Io ***@***.***> wrote:

> I compiles and works for me in Arduino for Teensy.
>
> Create *my_plugin.c* (not *my_plugin.c.txt* as sometimes Windows does -
> it has the IMO bad habit of hiding file extensions...) in the *src*
> folder, copy the code above into it and recompile. If you get compile
> errors then be sure you have an up to date grblHAL version, if you have
> post the compiler errors here.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/terjeio/ioSender/discussions/148#discussioncomment-1653078>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AM7D7UEUZKMJHFIBUTCY3CLUMKIR3ANCNFSM5HWP3KDA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>


-- 


Jonas Concepcion

// www.joncon.net

**Reply by terjeio** [reply]
You already have _my_plugin.c_ in the _src_ folder? If so then the code has to be merged with the one you already have - what does it contain?

If you mean the one in the _grbl_ folder then do not touch that, create a new as described.

**Reply by infamous-panda** [reply]
> You already have _my_plugin.c_ in the _src_ folder? If so then the code has to be merged with the one you already have - what does it contain?
> 
> If you mean the one in the _grbl_ folder then do not touch that, create a new as described.

Yes in the SRC folder I have the below

/*

  my_plugin.c - plugin template for setting auxillary output on feed hold

  Part of grblHAL

  Public domain.

*/

#include <string.h>

#include "driver.h"
#include "grbl/protocol.h"

static uint8_t port;

static on_state_change_ptr on_state_change;
static on_report_options_ptr on_report_options;

static void onStateChanged (sys_state_t state)
{
    static sys_state_t last_state = STATE_IDLE;

    if(state != last_state) {
        last_state = state;
        hal.port.digital_out(port, state == STATE_HOLD);
    }

    if(on_state_change)         // Call previous function in the chain.
        on_state_change(state);
}

static void onReportOptions (bool newopt)
{
    on_report_options(newopt);  // Call previous function in the chain.

    if(!newopt)                 // Add info about us to the $I report.
        hal.stream.write("[PLUGIN:MY PLUGIN Template 2]" ASCII_EOL);
}

static void output_warning (uint_fast16_t state)
{
    report_message("An output port is required for my_plugin!", Message_Warning);
}

// Tell the user about which port is used for the output
static void output_port (uint_fast16_t state)
{
    char msg[30];

    strcpy(msg, "My plugin port: ");
    strcat(msg, uitoa(port));

    report_message(msg, Message_Info);
}

void xmy_plugin_init()
{
    if(hal.port.num_digital_out == 0)                   // This plugin requires one digital output port,
        protocol_enqueue_rt_command(output_warning);    // complain if not available.

    else {
        port = hal.port.num_digital_out - 1;            // Claim the
        hal.port.num_digital_out--;                     // last free port.

        on_state_change = grbl.on_state_change;         // Subscribe to the state changed event by saving away the original
        grbl.on_state_change = onStateChanged;          // function pointer and adding ours to the chain.

        on_report_options = grbl.on_report_options;     // Add our plugin to to the options report chain
        grbl.on_report_options = onReportOptions;       // to tell the user we are active.

        protocol_enqueue_rt_command(output_port);       // Tell the user about which port is used for the output after startup is complete.
    }
}

**Reply by terjeio** [reply]
This is an experiment by you earlier? The code is inactive since the init function is named void `xmy_plugin_init()`, so IMO you can make a copy if you want and overwrite it.

**Comment by infamous-panda** [comment]
Yeah it doesn't work for me. My libraries are copied from Phil Barrett and he mentioned that they maybe a little older than the current. I had tried compiling my own from the latest files without luck so I asked Phil for his set. Error below

#include "grbl/hal.h"
#include "grbl/gcode.h"

static on_stream_changed_ptr on_stream_changed;
static on_report_options_ptr on_report_options;

static void stream_changed (stream_type_t type)
{
    if(on_stream_changed)
        on_stream_changed(type);

    if(type == StreamType_SDCard)
        gc_execute_block("M1");
}

static void report_options (bool newopt)
{
    on_report_options(newopt);

    if(!newopt)
        hal.stream.write("[PLUGIN:SD Pause v0.01]" ASCII_EOL);
}

void my_plugin_init (void)
{
    on_stream_changed = grbl.on_stream_changed;
    grbl.on_stream_changed = stream_changed;

    on_report_options = grbl.on_report_options;
    grbl.on_report_options = report_options;
}

**Reply by terjeio** [reply]
Try this hex:
[grblHAL_Teensy4_SD pause.ino.zip](https://github.com/terjeio/ioSender/files/7548989/grblHAL_Teensy4_SD.pause.ino.zip)

It would be great if we could sort out your compilation problems with the latest version.

**Reply by infamous-panda** [reply]
> Try this hex: [grblHAL_Teensy4_SD pause.ino.zip](https://github.com/terjeio/ioSender/files/7548989/grblHAL_Teensy4_SD.pause.ino.zip)
> 
> It would be great if we could sort out your compilation problems with the latest version.

Thank you, I was able to compile all from the most current posted. Things are working now. Is the improvement to have the gcode show up on the console from SD something to do with the firmware or the sender or both?

**Reply by terjeio** [reply]
Only the sender has to be modified if using FTP to download from the card. What is likely to be challenging is how to keep the program listing in sync with code run direcly from the card.
If download and run from the sender this will will be fairly easy to implement.

**Comment by infamous-panda** [comment]
Terjeio,

Your hex file works great. 

I got it to compile with updating the some of the libraries to the latest version, but the pause from SD-CARD does not work and it does not compile if I enable FTP. When I have time I will try to build out the library again with fresh files starting from scratch. But honestly I spent a lot of time on that process last week without success. 

I do not want to rely on your hex file for testing purposes as I cannot change parameters to experiment with. Maybe I will wait to look at this again when the SD files have a gcode readout like pc streaming files.
